internal/security/bootstrapper/command/setupacl/command.go

```diff
@@ -0,0 +1,364 @@
+		return interfaces.StatusCodeExitWithError, fmt.Errorf("failed to retrieve secretstore token: %v", err)
+	}
+
+	c.loggingClient.Info("successfully get secretstore token and configuring the registry access for secretestore")
+
+	// configure Consul access with both Vault mgmt token and consul's bootstrap token
+	if err := c.configureConsulAccess(secretstoreToken, bootstrapTokenInfo.SecretID); err != nil {
+		return interfaces.StatusCodeExitWithError, fmt.Errorf("failed to configure Consul access: %v", err)
+	}
+
+	return
+}
+
+// GetCommandName returns the name of this command
+func (c *cmd) GetCommandName() string {
+	return CommandName
+}
+
+func (c *cmd) getAPI_URL(api string) (string, error) {
+	apiURL := fmt.Sprintf("%s://%s:%d%s", c.configuration.StageGate.Registry.ACL.Protocol,
+		c.configuration.StageGate.Registry.Host, c.configuration.StageGate.Registry.Port, api)
+	_, err := url.Parse(apiURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse API URL: %v", err)
+	}
+	return apiURL, nil
+}
+
+// waitForNonEmptyConsulLeader is a special waitFor function on waiting for "non-empty" leader being available
+// the ordinary http waitFor won't work as the returned http status code from API call is 200 even when Consul's leader
+// is an empty string ("") but we need an non-empty leader; so 200 doesn't mean we have a leader
+func (c *cmd) waitForNonEmptyConsulLeader() error {
+	c.errs = make(chan error, 1)
+	// check Consul whether we have non-empty leader elected before proceed to Consul's ACL bootstrapping process
+	// as it requires non-empty leader to start with
+	go func() {
+		c.checkConsulLeader()
+		c.waitGroup.Wait()
+	}()
+
+	gotLeader := make(chan bool)
+	// a separate goroutine to keep waiting for a leader got elected from the checkConsulLeader goroutine
+	// until retryTimeout is reached otherwise
+	go func() {
+		for {
+			if err := <-c.errs; err != nil {
+				continue
+			}
+			gotLeader <- true
+			close(gotLeader)
+			return
+		}
+	}()
+
+	// block here until either we get a leader, or timeout is reached in which case ACL bootstrapping cannot continue
+	select {
+	case <-gotLeader:
+		c.loggingClient.Info("found Consul leader to bootstrap ACL")
+		return nil
+	case <-time.After(c.retryTimeout):
+		return errors.New("timed out to get non-empty Consul leader")
+	}
+}
+
+// checkConsulLeader is a goroutine constantly running if there is an error from API call
+func (c *cmd) checkConsulLeader() {
+	c.waitGroup.Add(1)
+	go func() {
+		defer c.waitGroup.Done()
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		// ticks every second to retry if we receive an error from API call
+		for range ticker.C {
+			if err := c.callConsulLeaderAPI(); err != nil {
+				c.loggingClient.Warnf("error from getting Consul leader API call, will retry it again: %v", err)
+				c.errs <- err
+				continue
+			}
+			// once reached here, we have a leader and terminate this goroutine
+			c.errs <- nil
+			return
+		}
+	}()
+}
```

Comment:
This code seems too complex for a simple retry loop. Why not just the code in line 206-218 w/o the channel for `waitForNonEmptyConsulLeader`

Reply:
i don't know how to do it just with those lines 206-218 and at the same time with time-out limit. let me know if you can figure out a better solution.

Reply:
Why? The timeout is there with the ticker?? 
Also I just noticed that this would need to return the timeout error after the for loop

Reply:
time out error is in **line 196-197**, we have already logged all the error during the getting leader process; there is no need for redundant error

Reply:
also, note that this function is go routine so no return for the function allowed actually

Reply:
Yes, a little refactoring is needed. Doesn't need to be a go func.

Reply:
let me know your solution, i really don't know how to do it gracefully without go func() when there are retries interval and time limit

Reply:
plus, to avoid data race condition if channels

Reply:
thanks for the help, looks great